pkg/readiness/objset.go

```diff
@@ -27,9 +27,42 @@ type objKey struct {
 	namespacedName types.NamespacedName
 }
 
-type objSet map[objKey]struct{}
-
 func (k objKey) String() string {
 	return fmt.Sprintf("%s [%s]", k.namespacedName.String(), k.gvk.String())
+}
+
+type objData struct {
+	retries int
+}
+
+// decrementRetries handles objData retries, and returns whether it's time to delete the objData.
+func (o *objData) decrementRetries() bool {
+	// if retries is less than 0, allowed retries are infinite
+	if o.retries < 0 {
+		return false
+	}
+
+	// If we have retries left, use one
+	if o.retries > 0 {
+		o.retries--
+		return false
+	}
 
+	// if we have zero retries, we can delete
+	return true
 }
+
+type objDataMutator func(*objData) *objData
+
+func setRetriesFromFlag(o *objData) *objData {
```

Comment:
`setRetriesFromFlag` and `mutatedObjData` could be:

```
func newObjDataFromFlags() *objData {
   return &objData{retries: *readinessRetries}
}
```

and for tests you could pass in the following function as an argument:

```
func() *objData { return &objData{ retries: <<whatever # of retries>> } }
```

This avoids having mutators embedded in the code, which is a bit more confusing to follow b/c of the extra layer of indirection.

Reply:
Changed to this implementation (which I called a Factory, since it generates a new instance of a type)

Why return a pointer?  It seems better to just return an `objData`.

Reply:
I was avoiding `Factory` because it seems like the cool golang kids don't use it (too Java), but the analogy is apt. Personally, I usually don't name the functional argument types unless the function signature is long or there is a lot of repetition, but both ways work.

WRT why return a pointer... I suppose it doesn't really matter either way, as this is a small struct that only returns an `int` and I don't really see us taking advantage of any mutate-in-place behavior.

I tend to default to struct pointers unless I'm returning something I'm persisting to a library consumer. Then I'll return a ref or a deepcopied pointer in order to protect the library's internal state.

TL;DR pointers can help for large structs, they may have a performance/memory penalty for smaller structs. Pointers vs structs also lead to shared state behavior that can be very useful (lots of mutate-in-place logic happening in mutation), or very annoying.

For an internal-to-the-package API I really don't put much thought into it b/c it's so cheap to change later.

Reply:
I commented in the review that I think the tip off here is that every place we're using the return value, we immediately dereference it.
As an aside, I like geeking out on the value vs pointer debate. I tend to default to values to avoid accidentally allowing mutations, but you can make bugs either way you cut it.
In terms of performance, pointers will escape to the heap and put more pressure on the GC, but also provide less data locality which leads to all kinds of cache line misses. Completely irrelevant micro-optimizations in our case :)

Reply:
Does golang make a heap vs. stack distinction for pointers vs. values? I would have thought the ability to send `&value` and have GC still work meant that they are treated the same WRT memory management?

There may be a golang subtlety I'm missing here though.

Reply:
Looks like there is a stack vs. heap optimization. I wonder when the decision is made... compile time?

https://medium.com/@meeusdylan/when-to-use-pointers-in-go-44c15fe04eac

Reply:
Yes, my understanding is that the compiler performs escape analysis, not during runtime. You can see some of the decisions it's making by passing `-gcflags '-m'` to your build.